modules/backendworker/backendworker.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/user"
+	backendscheduler_client "github.com/grafana/tempo/modules/backendscheduler/client"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/modules/storage"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	tempo_util "github.com/grafana/tempo/pkg/util"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+const (
+	reasonCompactorDiscardedSpans = "trace_too_large_to_compact"
+)
+
+type BackendWorker struct {
+	services.Service
+
+	cfg              Config
+	store            storage.Store
+	overrides        overrides.Interface
+	backendScheduler tempopb.BackendSchedulerClient
+
+	workerID string
+}
+
+// var tracer = otel.Tracer("modules/backendworker")
+
+func New(cfg Config, schedulerClientCfg backendscheduler_client.Config, store storage.Store, overrides overrides.Interface) (*BackendWorker, error) {
+	s := &BackendWorker{
+		cfg:       cfg,
+		store:     store,
+		overrides: overrides,
+	}
+
+	workerID, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	s.workerID = workerID
+
+	schedulerClient, err := backendscheduler_client.New(cfg.BackendSchedulerAddr, schedulerClientCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create backend scheduler client: %w", err)
+	}
+	s.backendScheduler = schedulerClient
+
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+	return s, nil
+}
+
+func (w *BackendWorker) starting(ctx context.Context) error {
+	if w.cfg.Poll {
+		w.store.EnablePolling(ctx, w)
+	}
+
+	return nil
+}
+
+func (w *BackendWorker) running(ctx context.Context) error {
+	// TODO: consider dropping the ticker and rely only on the backoff.
+	ticker := time.NewTicker(w.cfg.Interval)
+	defer ticker.Stop()
+
+	level.Info(log.Logger).Log("msg", "backend scheduler running")
+
+	b := backoff.New(ctx, w.cfg.Backoff)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			if err := w.processCompactionJobs(ctx); err != nil {
+				level.Error(log.Logger).Log("msg", "error processing compaction jobs", "err", err, "backoff", b.NextDelay())
+				b.Wait()
+				continue
+			}
+
+			b.Reset()
+		}
+	}
+}
+
+func (w *BackendWorker) processCompactionJobs(ctx context.Context) error {
+	// TODO: the org ID is not used by the backend scheduler, but it is required
+	// by the request.  Figure out how to disable this requirement.
+
+	// Request next job
+	resp, err := w.backendScheduler.Next(user.InjectOrgID(ctx, w.workerID), &tempopb.NextJobRequest{
+		WorkerId: w.workerID,
+		Type:     tempopb.JobType_JOB_TYPE_COMPACTION,
+	})
+	if err != nil {
+		return fmt.Errorf("error getting next job: %w", err)
+	}
+
+	if resp == nil || resp.JobId == "" {
+		return fmt.Errorf("no jobs available")
+	}
+
+	if resp.Detail.Tenant == "" {
+		return w.failJob(ctx, resp.JobId, "received job with empty tenant")
```

Comment:
it's a little odd to tell the scheduler a job is failed b/c it was a bad job, but i'm also not really sure what the best way to handle this edge case. it basically a suggests bug in the system.

i do think we need a metric for "asked for a job but didn't get one" that we alert on. i think "didn't get one" should be holistic and mean everything:

1) network call failed
2) "bad job" i.e. no tenant or other job invalid reasons
3) other crap

Reply:
I agree its a little strange.  Currently I don't see a way we end up with an empty tenant on the worker, but this was a little future proofing, since future code might push a job which has an empty tenant.  I've added a metric for this here, but open to other thoughts as we grow this functionality.